web_fragments/fragment.py

```diff
@@ -0,0 +1,271 @@
+"""
+Python representation of a web fragment.
+"""
+
+from collections import namedtuple
+
+FragmentResource = namedtuple("FragmentResource", "kind, data, mimetype, placement")  # pylint: disable=C0103
+
+JS_API_VERSION = 1
+
+
+class Fragment(object):
+    """
+    A fragment of a web page to be included on another page.
+
+    A fragment consists of HTML for the body of the page, and a series of
+    resources needed by the body. Resources are specified with a MIME type
+    (such as "application/javascript" or "text/css") that determines how they
+    are inserted into the page.  The resource is provided either as literal
+    text, or as a URL.  Text will be included on the page, wrapped
+    appropriately for the MIME type.  URLs will be used as-is on the page.
+
+    Resources are only inserted into the page once, even if many Fragments
+    in the page ask for them.  Determining duplicates is done by simple text
+    matching.
```

Comment:
This doesn't have to be a v1 thing, but we've been bitten performance-wise in the past by our really crude understanding of front end dependency loading. There's inline vs. linked, but there's also async (for JS, CSS in the body, fonts), fonts in general, web components (html imports), etc. Having an extensible way to load things via a smarter `FragmentResource` would be really helpful.